tests/core_tests/service_nodes.h

```diff
@@ -134,4 +134,41 @@ struct get_test_options<test_deregister_safety_buffer> {
   const cryptonote::test_options test_options = {
     hard_forks
   };
+};
+
+class test_deregisters_on_split : public test_chain_unit_base
+{
+
+  size_t m_invalid_tx_index;
+
+public:
+  test_deregisters_on_split();
+  bool generate(std::vector<test_event_entry>& events);
+  bool test_on_split(cryptonote::core& c, size_t ev_index, const std::vector<test_event_entry>& events);
+
+  bool mark_invalid_tx(cryptonote::core& /*c*/, size_t ev_index, const std::vector<test_event_entry>& /*events*/)
+  {
+    m_invalid_tx_index = ev_index + 1;
+    return true;
+  }
+
+  bool check_tx_verification_context(const cryptonote::tx_verification_context& tvc,
+                                     bool tx_added,
+                                     size_t event_idx,
+                                     const cryptonote::transaction& /*tx*/)
+  {
+    if (m_invalid_tx_index == event_idx)
+      return tvc.m_verifivation_failed;
+    else
+      return !tvc.m_verifivation_failed && tx_added;
+  }
+};
+
+template<>
+struct get_test_options<test_deregisters_on_split>
+{
+  const std::pair<uint8_t, uint64_t> hard_forks[3] = { std::make_pair(7, 0),
+                                                       std::make_pair(8, 1),
+                                                       std::make_pair(9, 2) };
```

Comment:
This causes an array overrun here:
https://github.com/loki-project/loki/blob/ffdf6826f77398520e69736032f2564d65afa9d5/src/cryptonote_core/blockchain.cpp#L336

Similarly for the other tests where you declare the hard-forks. The code in cryptonote_core iterates over hard_forks expecting that it is null terminated. So what hardforks should look like is 

```
const std::pair<uint8_t, uint64_t> hard_forks[4] = 
{ 
    std::make_pair(7, 0),
    std::make_pair(8, 1),
    std::make_pair(9, 2),
    std::make_pair(0, 0),
};
```

I think there's a good case in making cryptonote::test_options have int num_hard_forks, instead of the implicit knowledge, this has to be null terminated.

Reply:
Do we have an update on this?